server/src/main/java/org/opensearch/action/pagination/ShardPaginationStrategy.java

```diff
@@ -0,0 +1,273 @@
+    ) {
+        if (Objects.isNull(lastIndexName) || Objects.isNull(lastIndexCreationTime)) {
+            return PaginationStrategy.getSortedIndexMetadata(
+                clusterState,
+                PageParams.PARAM_ASC_SORT_VALUE.equals(sortOrder) ? ASC_COMPARATOR : DESC_COMPARATOR
+            );
+        } else {
+            return PaginationStrategy.getSortedIndexMetadata(
+                clusterState,
+                getMetadataFilter(sortOrder, lastIndexName, lastIndexCreationTime),
+                PageParams.PARAM_ASC_SORT_VALUE.equals(sortOrder) ? ASC_COMPARATOR : DESC_COMPARATOR
+            );
+        }
+    }
+
+    private static Predicate<IndexMetadata> getMetadataFilter(String sortOrder, String lastIndexName, Long lastIndexCreationTime) {
+        if (Objects.isNull(lastIndexName) || Objects.isNull(lastIndexCreationTime)) {
+            return indexMetadata -> true;
+        }
+        return indexNameFilter(lastIndexName).or(
+            IndexPaginationStrategy.getIndexCreateTimeFilter(sortOrder, lastIndexName, lastIndexCreationTime)
+        );
+    }
+
+    private static Predicate<IndexMetadata> indexNameFilter(String lastIndexName) {
+        return metadata -> metadata.getIndex().getName().equals(lastIndexName);
+    }
+
+    /**
+     * Will be used to get the list of shards and respective indices to which they belong,
+     * which are to be displayed in a page.
+     * Note: All shards for a shardID will always be present in the same page.
+     */
+    private PageData getPageData(
+        List<IndexMetadata> filteredIndices,
+        Map<String, IndexRoutingTable> indicesRouting,
+        final ShardStrategyToken token,
+        final int numShardsRequired
+    ) {
+        List<ShardRouting> shardRoutings = new ArrayList<>();
+        List<String> indices = new ArrayList<>();
+        int shardCount = 0;
+        IndexMetadata lastAddedIndex = null;
+
+        // iterate over indices until shardCount is less than numShardsRequired
+        for (IndexMetadata indexMetadata : filteredIndices) {
+            String indexName = indexMetadata.getIndex().getName();
+            boolean indexShardsAdded = false;
+            // Always start from shardID 0 for all indices except for the first one which might be same as the last sent
+            // index. To identify if an index is same as last sent index, verify both the index name and creaton time.
+            int startShardId = shardCount == 0 ? getStartShardIdForPageIndex(token, indexName, indexMetadata.getCreationDate()) : 0;
+            Map<Integer, IndexShardRoutingTable> indexShardRoutingTable = indicesRouting.get(indexName).getShards();
+            for (; startShardId < indexShardRoutingTable.size(); startShardId++) {
+                if (indexShardRoutingTable.get(startShardId).size() > numShardsRequired) {
+                    throw new IllegalArgumentException("size value should be greater than the replica count of all indices");
```

Comment:
when will this exception be thrown, can you give an example? what size is this?

Reply:
> what size is this?


This is user provided pageParam ```size```.


> when will this exception be thrown, can you give an example?

Its for the case when ShardPaginationStrategy is provided a ```pageSize/size``` value which is smaller than any of the index's replica count. Since, the ```shardRoutings/replicas``` of a ```shardID``` cannot get split across pages, the size value should always be greater than the replica count.


Conversation [Link](https://github.com/opensearch-project/OpenSearch/pull/14641#discussion_r1777420857)

Reply:
you have min size check of 1000. is this still needed?

Reply:
update the error message to ```size value should be greater than the replica count across all indices, so that all primary and replicas of a shard show up in single page```

Reply:
I think it's better to have this check, so that it stays decoupled from Rest layer. If in future we choose to use this for any other purpose, it would help the users to explicitly take a note of this functionality which otherwise could get overlooked.

Reply:
Updated the exception message to:
```
size value should be greater than the replica count across all indices, so that all primary and replicas of a shard show up in single page

```